### PR TITLE
Cancelling all pending messages support

### PR DESCRIPTION
**Pull request by balsmn**
`CaliforniumLwM2mRequestSender `implementation had this method (cancelPendingMessages) and this concrete implementation was directly used in LeshanServer. This prevented from using the new `LwM2mRequestSender` API (containing support for both queue and non-queue mode) in Leshan Server. This PR aims to add the `cancelPendingMessages` to `LwM2mRequestSender` API and implements the same in all the concrete implementations (`LwM2mRequestSenderImpl`, `QueuedRequestSender` and `CaliforniumLwM2mRequestSender`). With this PR, Leshan Server will officially have the QueueMode support.

Signed-off-by: Bala Azhagappan <balasubramanian.azhagappan@bosch-si.com>

### CODE REVIEW

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java [hunk 1]
```diff
@@ -47,29 +47,30 @@
...
+            LwM2mModelProvider tModelProvider, LwM2mRequestSender tLwM2mRequestSender,
+            MessageStore tInMemoryMessageStore) {
+
+        this.coapServer = tCoapServer;
+        this.clientRegistry = tClientRegistry;
+        this.observationRegistry = tObservationRegistry;
+        this.securityRegistry = tSecurityRegistry;
+        this.modelProvider = tModelProvider;
+        this.lwM2mRequestSender = tLwM2mRequestSender;
+        this.messageStore = tInMemoryMessageStore;
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java]: (just cosmetic, but I don't really like this 't' ? any good reason for this change ?)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/LwM2mRequestSenderImpl.java [hunk 2]
```diff
@@ -94,4 +94,13 @@ public void stop() {
...
     }
+
+    @Override
+    public void cancelPendingRequests(Client client) {
+        if (client.usesQueueMode()) {
+            queuedRequestSender.cancelPendingRequests(client);
+        } else {
+            defaultRequestSender.cancelPendingRequests(client);
+        }
+    }
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/LwM2mRequestSenderImpl.java]: You must cancel request on both 2 senders because, for the same registration you could have pending request in queuemode and not queuemode (as binding mode can change)
(I agree this Far-fetched but I see no problem to just call `cancelPendingRequests` on both sender)

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/MessageStore.java [hunk 3]
```diff
@@ -55,16 +57,16 @@
...
      * @param endpoint client endpoint name
      */
     void deleteFirst(String endpoint);
+
+    /**
+     * 
+     * @param endpoint
+     * @return
+     */
+    List<QueuedRequest> removeAll(String endpoint);
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/MessageStore.java]: why removing javadoc ?

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2015 Bosch Software Innovations GmbH and others.
+ * Copyright (c) 2016 Bosch Software Innovations GmbH and others.
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java]: Please avoid this kind of changes (just changing formatting or copyright year), all this changes is not linked to the PR. This make the review harder.

**Comment by sbernard31** [comment]
I do not check the unit test part in details.
(I will not be available the next 2 weeks)

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java (hunk 1 above)
Comment (balsmn) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java]: yes. Otherwise to access member variable one has to use for ex. `QueueModeLeshanServer.this.observationRegistry`

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/LwM2mRequestSenderImpl.java (hunk 2 above)
Comment (balsmn) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/LwM2mRequestSenderImpl.java]: Should not be a problem as `QueuedRequestSender` has its own instance of `CaliforniumLwM2mRequestSender`.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/MessageStore.java (hunk 3 above)
Comment (balsmn) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/MessageStore.java]: Actually javadoc is not complete here. I removed this method first and again added it with a different signature.

**Comment by balsmn** [comment]
can then any of the others (@msangoi @sophokles73 @jvermillard ) look at this PR and bring it to a close?

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java (hunk 4 above)
Comment (balsmn) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java]: I agree it makes the review harder. Before commiting, took a a look at the previous PR's regarding this and followed the same like in PR #150.

Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java]: In this PR, this is new interfaces. (but as they have the same name as previous classes, so git diff show this like a modification)
(And the change is done a separate commit to facilitate review)
Do not hesitate to comment my PR, if you see something wrong :smile:

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java (hunk 1 above)
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java]: Personnaly, I prefer QueueModeLeshanServer.this.observationRegistry.
(the "t" stand for what ?)

**Comment by sbernard31** [comment]
AFAIK, the whole team will be unavailable for at least one week.
You are alone on this one ;)
alone

(I think the PR is globally good so you can continue on this base for the next step, thx @balsmn )

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java (hunk 1 above)
Comment (balsmn) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/util/QueueModeLeshanServer.java]: It is a short form for "the". At first look I was a bit confused with the syntax `QueueModeLeshanServer.this.observationRegistry`. But I am not so religious about this and will revert it back.

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java (hunk 4 above)
Comment (balsmn) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/QueueModeIntegrationTestHelper.java]: Yes, will do and I will keep this noted for future PR's. :-)

**Comment by balsmn** [comment]
can we bring this PR to a close?

**Comment by sbernard31** [comment]
This is now integrated in master (commit cd6738ca0933f120e25ff1e703d2925f73a766ce)
